src/psyclone/psyGen.py

```diff
@@ -362,8 +371,80 @@ def get(self, invoke_name):
                                       str(self.names)))
 
     def gen_code(self, parent):
+        '''
+        Create the f2pygen AST for each Invoke in the PSy layer.
+
+        :param parent: the parent node in the AST to which to add content.
+        :type parent: `psyclone.f2pygen.ModuleGen`
+        '''
+        opencl_kernels = []
         for invoke in self.invoke_list:
             invoke.gen_code(parent)
+            # If we are generating OpenCL for an Invoke then we need to
+            # create routine(s) to set the arguments of the Kernel(s) it
+            # calls. We do it here as this enables us to prevent
+            # duplication.
+            if invoke.schedule.opencl:
+                for kern in invoke.schedule.kern_calls():
+                    if kern.name not in opencl_kernels:
+                        opencl_kernels.append(kern.name)
+                        kern.gen_arg_setter_code(parent)
+                # We must also ensure that we have a kernel object for
+                # each kernel called from the PSy layer
+                self.gen_ocl_init(parent, opencl_kernels)
+
+    def gen_ocl_init(self, parent, kernels):
```

Comment:
Could you make this static. pylint doesn't like it.

Reply:
Done.